[#4491] Implement support for OS-native configurations places

cleac opened the pull request:
**What this PR does / why we need it**:

This PR implements possibility for helm to store configuration at places, where they are expected to see for different operating systems including [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) and native for Windows and OSx places using [xdg-go](github.com/casimir/xdg-go) library.

Implementation will keep backward compatibility with previous version location of helm configurations (`$HOME/helm`, will be called "old" further), but will print warnings when it is used. Newer (one that is proposed) place is in higher priority, and it will use it if won't find an "old" one. 

Closes #4491

**Special notes for your reviewer**:

~This is a proposal, I'm going to test this during few hours, just to be sure it works as I expect.~

**If applicable**:
- [x] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

cleac committed:
[#4491] Implement support for XDG Base Dir Specs

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

cleac committed:
Update docs, add warning and install xdg-go package

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

bacongobbler commented on docs/install_faq.md:
```diff
@@ -104,5 +104,9 @@ I am trying to remove stuff.
-located by default in `~/.helm`.
+Along with the `helm` binary, Helm stores some files in `$HELM_HOME`. Check those
+locations to remove all the configurations too:
+ - `$XDG_CONFIG_DIR/helm` on *nix;
```
can we list out what $XDG_CONFIG_DIR is on Linux (`$HOME/.config/helm`)?

bacongobbler replied:
Also for consistency, can we change *nix to Linux? We refer to "anything other than Windows/MacOS" as linux in other parts of the documentation.

cleac replied:
$XDG_CONFIG_DIR is different on different unix-based operating system, so it is correct to say this just for linux. And I there are build for different *nix platforms than Linux. Will improve this

bacongobbler commented on cmd/helm/root.go:
```diff
@@ -42,7 +42,11 @@ Common actions from this point include:
 Environment:
-  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in ~/.helm
+  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in 
+					  "$XDG_CONFIG_DIR/helm"" on *nix, "%APPDATA%\helm" on Windows and 
```
can we list out what $XDG_CONFIG_DIR is on Linux? Something like

```
$HELM_HOME          set an alternative location for Helm files. By default, these are stored in 
					  "$XDG_CONFIG_DIR/helm" (typically $HOME/.config/helm) on *nix, "%APPDATA%\helm" on Windows and
```

Also there's an extra double quote after `$XDG_CONFIG_DIR/helm`. Would you mind removing that please?

cleac replied:
Fixed :smile:

bacongobbler commented on cmd/helm/init.go:
```diff
@@ -31,7 +31,7 @@ import (
 
 const initDesc = `
-This command sets up local configuration in $HELM_HOME (default ~/.helm/).
+This command sets up local configuration in $HELM_HOME (default $XDG_CONFIG_DIR/helm/).
```
same here; can we list out what exactly is usually $XDG_CONFIG_DIR? Many people know about $HOME, but not much about $XDG_CONFIG_DIR.

> This command sets up local configuration in $HELM_HOME (default $XDG_CONFIG_DIR/helm/). Based on your OS, $XDG_CONFIG_DIR is 

- `%APPDATA%\helm` on Windows
- `$HOME/Library/Preferences` on OSX
- `$HOME/.config/helm` on Linux

cleac replied:
Updated and fixed according to requests

bacongobbler commented on pkg/helm/environment/environment.go:
```diff
@@ -32,8 +34,21 @@ import (
+	} else if _, err := os.Stat(oldDefaultHelmHome); err != nil {
+		return defaultHelmHome
+	}
+	fmt.Printf("WARNING: using old-style configuration directiry. Please, consider moving it to %s\n", defaultHelmHome)
```
It might be better if this warning was moved to the `cmd` package anyways, as calling `fmt.Fprintf` here doesn't allow us to print it on different outputs (to a file, stderr, etc).

bacongobbler reviewed (changes requested):
Thanks for the PR! There's a few changes that need to happen before this lands, but this looks like a good start. Thanks for taking the time to PR this!

cleac committed:
Decouple determining home dir to custom directory and add tests.

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

cleac commented:
@bacongobbler thanks very much for a review, I've made some changes to improve PR. 

I've made some tests although I am afraid it should be improved, feels a bit strange, could you please give a hint if it could be implemented a bit cleaner from Go perspective? Also, the library logic of scope is already tested there, so there is no need to test it on helm side.

adamreese reviewed (changes requested):
Thanks for submitting this!

The xdg library does not respect `XDG_CONFIG_HOME` on darwin.  The precedence should be:

```
    1. Flag `--home`
    2. Environment set `HELM_HOME`
    3. Environment set `XDG_CONFIG_HOME`
    4. OS default
    5. HOME/.helm
```

cleac commented on pkg/helm/helmpath/default_home_dir.go:
```diff
@@ -0,0 +1,63 @@
+
+// Check whether new default helm home exists
+// TODO: improve me
+var DefaultHelmHomeExists = func() bool {
```
It’s for easier mocking these functions in tests. If there is better way, can you give me a clue, ‘cause I don’t like this way :(

cleac replied:
Improved this in new commit, can you look at it please?  :smile:

bacongobbler commented on pkg/helm/helmpath/default_home_dir.go:
```diff
@@ -0,0 +1,63 @@
+	if DefaultHelmHomeExists() || !OldDefaultHelmHomeExists() {
+		return defaultHelmHome
+	}
+	fmt.Fprintf(out, "WARNING: using old-style configuration directory. Please, consider moving it to %s\n", defaultHelmHome)
```
Yes, that's fine. We want to accept both ~/.helm and ~/.config/helm as valid locations.

cleac replied:
Ok, will do :smile:

cleac commented:
@adamreese as far as I inspected the code, it should respect the XDG_* variables... Did you test this one?

cleac commented:
> Thanks for submitting this!
> 
> The xdg library does not respect `XDG_CONFIG_HOME` on darwin. The precedence should be:
> 
> ```
>     1. Flag `--home`
>     2. Environment set `HELM_HOME`
>     3. Environment set `XDG_CONFIG_HOME`
>     4. OS default
>     5. HOME/.helm
> ```

@adamreese as far as I inspected the code, it should respect the XDG_* variables... Did you test this one?

cleac committed:
Improve implementation

Make test be better designed to use interface implementation instead
of leaking variables with closures. The removed warning was concidered
as irrelevant, so it won't be needed in fact, because legacy version
is not goint to become deprecated

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

bacongobbler commented:
I just tested this and confirmed the same behaviour @adamreese suggested.

```
><> ./bin/helm init
Creating /Users/bacongobbler/Library/Preferences/helm
Creating /Users/bacongobbler/Library/Preferences/helm/repository
Creating /Users/bacongobbler/Library/Preferences/helm/repository/cache
Creating /Users/bacongobbler/Library/Preferences/helm/plugins
Creating /Users/bacongobbler/Library/Preferences/helm/starters
Creating /Users/bacongobbler/Library/Preferences/helm/cache/archive
Creating /Users/bacongobbler/Library/Preferences/helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
$HELM_HOME has been configured at /Users/bacongobbler/Library/Preferences/helm.
Happy Helming!
><> XDG_CONFIG_HOME=~/.config ./bin/helm init
$HELM_HOME has been configured at /Users/bacongobbler/Library/Preferences/helm.
Happy Helming!
><> export XDG_CONFIG_HOME=~/.config
><> ./bin/helm init
$HELM_HOME has been configured at /Users/bacongobbler/Library/Preferences/helm.
Happy Helming!
```

Even removing the directory and trying again didn't seem to work

```
><> rm -rf /Users/bacongobbler/Library/Preferences/helm
><> ./bin/helm init
Creating /Users/bacongobbler/Library/Preferences/helm
Creating /Users/bacongobbler/Library/Preferences/helm/repository
Creating /Users/bacongobbler/Library/Preferences/helm/repository/cache
Creating /Users/bacongobbler/Library/Preferences/helm/plugins
Creating /Users/bacongobbler/Library/Preferences/helm/starters
Creating /Users/bacongobbler/Library/Preferences/helm/cache/archive
Creating /Users/bacongobbler/Library/Preferences/helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
$HELM_HOME has been configured at /Users/bacongobbler/Library/Preferences/helm.
Happy Helming!
```

So this PR doesn't appear to respect when I change XDG_CONFIG_HOME.

bacongobbler commented:
HELM_HOME and `--home` still seems to work though.

```
><> HELM_HOME=~/.config/helm ./bin/helm init
Creating /Users/bacongobbler/.config/helm
Creating /Users/bacongobbler/.config/helm/repository
Creating /Users/bacongobbler/.config/helm/repository/cache
Creating /Users/bacongobbler/.config/helm/plugins
Creating /Users/bacongobbler/.config/helm/starters
Creating /Users/bacongobbler/.config/helm/cache/archive
Creating /Users/bacongobbler/.config/helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
$HELM_HOME has been configured at /Users/bacongobbler/.config/helm.
Happy Helming!
```

cleac commented:
Huuuh, that feels very awkward :thinking: Thanks for pointing that, will check this one deeper

ian-howell commented on pkg/helm/helmpath/default_home_dir.go:
```diff
@@ -0,0 +1,65 @@
+//	- %APPDATA%\helm on Windows
+//	- ~/Library/Preferences/helm on OSX
+//  - $XDG_CONFIG_DIR/helm (typically ~/.config/helm for linux)
+var defaultHelmHome = filepath.Join(xdg.ConfigHome(), "helm")
```
I think this is what you want

```
var defaultHelmHome = xdg.App{Name: "helm"}.ConfigPath("")
```

ian-howell commented:
Just tested with my suggested change. Seems to be working as long as `$HELM_HOME` isn't set, but I think that's expected behavior
```
➜  helm git:(xdg) XDG_CONFIG_HOME=~/.config ./bin/helm init
Creating /home/ian/.config/helm 
Creating /home/ian/.config/helm/repository 
Creating /home/ian/.config/helm/repository/cache 
Creating /home/ian/.config/helm/plugins 
Creating /home/ian/.config/helm/starters 
Creating /home/ian/.config/helm/cache/archive 
Creating /home/ian/.config/helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
$HELM_HOME has been configured at /home/ian/.config/helm.
Happy Helming!
➜  helm git:(xdg) rm ~/.config/helm -r
➜  helm git:(xdg) export XDG_CONFIG_HOME=~/.config
➜  helm git:(xdg) ./bin/helm init
Creating /home/ian/.config/helm 
Creating /home/ian/.config/helm/repository 
Creating /home/ian/.config/helm/repository/cache 
Creating /home/ian/.config/helm/plugins 
Creating /home/ian/.config/helm/starters 
Creating /home/ian/.config/helm/cache/archive 
Creating /home/ian/.config/helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
$HELM_HOME has been configured at /home/ian/.config/helm.
Happy Helming!

```

bacongobbler commented:
Awesome. Thanks for finding the issue, @ian-howell! @cleac if you wouldn't mind making that change, this should work. We should also rebase this against master to get the fixes implemented in #4934 so that appveyor CI tests can check that this PR works on Windows.

cleac commented:
@ian-howell thanks much for pointing this out! I got confused with this kind of issue a bit :sweat_smile: 

@bacongobbler I will   update my PR with this fix applied and rebased upon dev-v3 during 10-15 minutes

cleac committed:
Fix using custom XDG_CONFIG_HOME if available

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

cleac commented:
Tested this one out and it seems to work as well. It feels pretty strange, why it couldn't work the way I used it, but this is how things work.

ian-howell commented on cmd/helm/root_test.go:
```diff
@@ -17,8 +17,8 @@ limitations under the License.
-	"path/filepath"
 	"testing"
 
 	"k8s.io/client-go/util/homedir"
```
One more thing, the CI gate is failing because "k8s.io/client-go/util/homedir" is no longer used. You'll need to delete this line.

Also, thanks for the quick response and fix :smile:

cleac replied:
Fixed this one too :) Thank you :D

cleac committed:
Remove redundant dependency

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

bacongobbler commented:
I'll give @adamreese a few days to check this one over, then we should be good to merge.

bacongobbler commented on pkg/helm/helmpath/default_home_dir.go:
```diff
@@ -0,0 +1,65 @@
+//	- %APPDATA%\helm on Windows
+//	- ~/Library/Preferences/helm on OSX
+//  - $XDG_CONFIG_DIR/helm (typically ~/.config/helm for linux)
+var defaultHelmHome = xdg.App{Name: "helm"}.ConfigPath("")
```
`pkg/helmpath` is purposely built around path traversal within `helm home`, but isn't intended to determine *where* `helm home` lives. Can you please move all of the logic surrounding `defaultHelmHome` out of `pkg/helmpath` and into `pkg/helm/environment`? Thanks.

bacongobbler commented:
Still doesn't seem to work in certain scenarios either.

```
><> XDG_CONFIG_HOME=/tmp helm home
/home/bacongobbler/.helm
```

ian-howell commented:
> Still doesn't seem to work in certain scenarios either. Can you please look into this and make sure it works?
> 
> ```
> ><> XDG_CONFIG_HOME=/tmp helm home
> /home/bacongobbler/.helm
> ```

Won't you need to export `XDG_CONFIG_HOME` for helm to see it?
```
➜  helm git:(xdg) ./bin/helm home
/home/ian/.config/helm
➜  helm git:(xdg) XDG_CONFIG_HOME=~/xdg-test ./bin/helm home
/home/ian/xdg-test/helm
➜  helm git:(xdg) export XDG_CONFIG_HOME=~/xdg-test
➜  helm git:(xdg) ./bin/helm home
/home/ian/xdg-test/helm
```

ian-howell committed:
Move home dir logic to package environment

Signed-off-by: Ian Howell <ian.howell0@gmail.com>

ian-howell commented:
@cleac Pull request here for @bacongobbler's requested changes: https://github.com/cleac/helm/pull/1

bacongobbler commented:
> Won't you need to export `XDG_CONFIG_HOME` for helm to see it?

I think it has something to do with how xdg-go exaluates environment variables at runtime. For example, `HELM_HOME=/foo helm home` works just fine. Setting XDG_CONFIG_HOME on-the-fly should work as well.

From the tip of dev-v3:

```
><> HELM_HOME=/tmp ./bin/helm home
/tmp
```

ian-howell commented:
Right, but that's because when you prefix a line with a variable declaration, [that variable is exported to the command](https://unix.stackexchange.com/a/126942). My last comment demonstrates the same behavior for `XDG_CONFIG_HOME` in the second line.

In other words,
```
➜  helm git:(dev-v3) HELM_HOME=/tmp
➜  helm git:(dev-v3) ./bin/helm home
/home/ian/.helm
```

(I really hope I'm not sounding snippy btw, I'm just pretty sure that this implementation works :relaxed:)

bacongobbler commented:
The key point here is

> when you prefix a line with a variable declaration, **that variable is exported to the command**.

So when I do `HELM_HOME=tmp helm home`, HELM_HOME is exported by the shell and viewed by Helm as an environment variable. As such, I should expect that command should set `helm home` to `/tmp`.

You can even try it with other commands:

```
><> HELM_HOME=/tmp env | grep HELM
HELM_HOME=/tmp
```

In other words, the example I provided above where we are calling `XDG_CONFIG_HOME=/foo helm home` *is* injecting XDG_CONFIG_HOME into the environment. The PR is just not picking it up and respecting it.

We had issues long ago with race conditions where environment variables weren't being picked up during runtime, so this isn't the first time we've seen this come up. See #2476 and #2673 for more context. Hope this helps.

ian-howell commented:
Right, I misread one of your comments, my bad :flushed:
So I did some experimentation. It seems that the PR *is* respecting the variable, but the order of precedence is not correct. It appears to currently use this precedence:
```
    1. Flag `--home`
    2. Environment set `HELM_HOME`
    3. HOME/.helm (if it exists)
    4. Environment set `XDG_CONFIG_HOME`
    5. OS default
```
I'm going to poke at this and see if I can't figure it out

ian-howell commented:
So the precedence I mentioned earlier actually *is* the desired behavior, as it maintains backward compatibility. Per the original message for this pull request:

> Implementation will keep backward compatibility with previous version location of helm configurations (`$HOME/helm`, will be called "old" further), but will print warnings when it is used. Newer (one that is proposed) place is in higher priority, and it will use it if won't find an "old" one.

Albeit, no message is printed. And further, perhaps this is not what the Helm community wants. But currently, this PR does exactly what it says it should do.

bacongobbler commented:
Hmm, I'm not sure what changed since the last time I tested this, but it seems to work now. Perhaps I did have an old ~/.helm still hanging around. Sorry for the trouble.

```
><> rm -rf ~/.config/helm/
><> rm -rf ~/.helm/
><> XDG_CONFIG_HOME=/tmp ./bin/helm init
Creating /tmp/helm 
Creating /tmp/helm/repository 
Creating /tmp/helm/repository/cache 
Creating /tmp/helm/plugins 
Creating /tmp/helm/starters 
Creating /tmp/helm/cache/archive 
Creating /tmp/helm/repository/repositories.yaml 
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com 
$HELM_HOME has been configured at /tmp/helm.
Happy Helming!
><> XDG_CONFIG_HOME=/tmp ./bin/helm home
/tmp/helm
```

Looks like this still needs a rebase before it can be merged. I'll take another look at the code :)

bacongobbler commented on docs/install_faq.md:
```diff
@@ -104,5 +104,9 @@ I am trying to remove stuff.
-located by default in `~/.helm`.
+Along with the `helm` binary, Helm stores some files in `$HELM_HOME`. Check those
+locations to remove all the configurations too:
+ - `$XDG_CONFIG_DIR/helm` (typically ~/.config/helm) on Linux;
```
Same here.

cleac commented on cmd/helm/root.go:
```diff
@@ -42,7 +42,10 @@ Common actions from this point include:
 
 Environment:
-  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in ~/.helm
+  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in
```
I think `NOTE` is pretty relevant to inform users that there is no need move directory somewhere. Maybe this should be added to `helm init` though?

cleac committed:
Reword a bit docs and make more consistent and clear

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

cleac commented:
Improved docs a bit :smile: (sorry it took so long)

Also, regarding issue with overriding default helm home on XDG_CONFIG_HOME - is it resolved? As far as I read the thread, it looks like it does, but I'm not sure.

ian-howell committed:
Fix a dependency issue

Signed-off-by: Ian Howell <ian.howell0@gmail.com>

ian-howell committed:
Clear up documentation related to XDG_CONFIG_HOME location

Signed-off-by: Ian Howell <ian.howell0@gmail.com>

ian-howell committed:
Add comments to appease the linter

Signed-off-by: Ian Howell <ian.howell0@gmail.com>

bacongobbler commented:
hey @cleac, yes it looks like the issues have been resolved. I'll give this another review tomorrow. Thanks for following up!

bacongobbler commented on cmd/helm/init.go:
```diff
@@ -36,7 +36,10 @@ import (
+This command sets up local configuration in $XDG_CONFIG_HOME/helm
+($HOME/.config/helm on Linux, %APPDATA%\helm on Windows and
+$HOME/Library/Preferences on OSX).
+NOTE: if there is old-style "$HOME/.helm" directory, it will be preffered.
```
Would you mind changing this to:

> NOTE: if $HOME/.helm exists, it will be used instead of $XDG_CONFIG_HOME/helm.

adamreese replied:
Also `~/.helm` not be used during `helm init`.  Init should always create $HELM_HOME || $XDG_CONFIG_HOME/helm.  `~/.helm` should only be used as a fallback for lookup.

bacongobbler commented on cmd/helm/root.go:
```diff
@@ -44,10 +44,13 @@ Common actions from this point include:
+  $HELM_HOME          set an alternative location for Helm files. By default, these are stored in
+                      "$XDG_CONFIG_DIR/helm" ("$HOME/.config/helm" on Linux,
+                      "%APPDATA%\helm" on Windows and "$HOME/Library/Preferences" on OSX).
+                      NOTE: if there is old-style "$HOME/.helm" directory, it will be used.
```
Can you please remove the quotes around the environment variables?

```
$HELM_HOME          set an alternative location for Helm files. By default, these are stored in
                      $XDG_CONFIG_DIR/helm ($HOME/.config/helm on Linux,
                      %APPDATA%\helm on Windows and $HOME/Library/Preferences on OSX).
                      NOTE: if $HOME/.helm exists, it will be used instead of $XDG_CONFIG_HOME/helm.
```

bacongobbler commented on docs/install_faq.md:
```diff
@@ -104,5 +104,9 @@ I am trying to remove stuff.
+ - `$XDG_CONFIG_HOME/helm` (typically `$HOME/.config/helm`) on Linux;
+ - `%APPDATA%\helm` on Windows;
+ - `$HOME/Library/Preferences` on OSX;
+ - `$HOME/.helm` if you were using old version of helm previously.
```
Please change this to:

> Along with the `helm` binary, Helm stores some files in `$HELM_HOME`. Check the following locations:
> - `$XDG_CONFIG_HOME/helm` (typically `$HOME/.config/helm`) on Linux;
> - `%APPDATA%\helm` on Windows;
> - `$HOME/Library/Preferences` on OSX;
> - `$HOME/.helm` if you were using an older version of Helm.

bacongobbler reviewed (commented):
Mostly doc changes at this point. Once those are done I think this is ready to merge. Thanks for taking the time to write this!

bacongobbler commented on cmd/helm/init.go:
```diff
@@ -36,7 +36,10 @@ import (
-This command sets up local configuration in $HELM_HOME (default ~/.helm/).
+This command sets up local configuration in $XDG_CONFIG_HOME/helm
+($HOME/.config/helm on Linux, %APPDATA%\helm on Windows and
+$HOME/Library/Preferences on OSX).
```
Should be `$HOME/Library/Preferences/helm` (missing `/helm` at the end)

ian-howell commented:
@cleac are you planning on finishing this? I'll gladly pick up the rest of the work if not :)

cleac commented:
@ian-howell will do during hour or two. Have been out for some time :smile:

cleac committed:
Update docs and add one missing place

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

cleac commented:
Have updated a PR, would be happy for improvement suggestions :smile:

ian-howell commented on cmd/helm/root.go:
```diff
@@ -45,12 +45,12 @@ Common actions from this point include:
-                      NOTE: if there is old-style "$HOME/.helm" directory, it will be used.
+                      $XDG_CONFIG_DIR/helm ($HOME/.config/helm on Linux, %APPDATA%\helm on Windows
+                      and $HOME/Library/Preferences/helm on macOS).
+                      NOTE: if there is old-style $HOME/.helm directory, it will be used.
```
I think this would read better like this:
```
NOTE: if $HOME/.helm exists, it will be used instead of $XDG_CONFIG_HOME/helm.
```

cleac replied:
Yeah, this looks better

cleac replied:
On the other hand, it doesn't describe the behavior as it is. The more accurate description of behavior should be like this:
```
NOTE: if default does not exist but $HOME/.helm does, the last will be used.
```

What do you think about this?

bacongobbler replied:
Made a similar comment in a previous review: https://github.com/helm/helm/pull/5166#discussion_r263210600

cleac replied:
@bacongobbler 
That comment stills not close to how things work :confused: This one describes it more accurately in fact:
```
NOTE: if default does not exist but $HOME/.helm does, the last will be used.
```

ian-howell replied:
@cleas After reviewing the code again, I see what you're saying, and I agree, though I think it could use just a little tweak:

NOTE: if **the default directory** does not exist but $HOME/.helm does, the **latter** will be used.

ian-howell commented on docs/faq.md:
```diff
@@ -44,8 +44,11 @@ with `helm repo add...`.
+ * `$HOME/.config/helm` on Linux
+ * `$HOME/Library/Preferences/helm` on macOS
+ * `%APPDATA%\helm` on Windows
+ * `$HOME/.helm` if you have used helm before 3rd release
```
Could you change this as well:

```
`$HOME/.helm` if it exists. This happens if you have used an older version of Helm
```

ian-howell commented on docs/glossary.md:
```diff
@@ -57,7 +57,11 @@ capitalized. When speaking of the client, _helm_ is in lowercase.
+ * `$HOME/.config/helm` on Linux
+ * `$HOME/Library/Preferences/helm` on OSx
+ * `%APPDATA%\helm` on Windows
+ * `$HOME/.helm` if you have used helm before 3rd release
```
Same as above

ian-howell reviewed (changes requested):
I'm not sure what changed to cause failures, but this PR is no longer passing the Circle CI gate. You probably want to be running a `make test` before pushing.
You can probably ignore the Appveyor issue, that gate will be removed soon.

cleac committed:
Small docs improvements and fix lints

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

ian-howell commented on pkg/helm/environment/default_home_dir.go:
```diff
@@ -0,0 +1,70 @@
+
+// GetDefaultConfigHome determines the configuration home dir.
+func GetDefaultConfigHome() string {
+	if ConfigPath.xdgHomeExists() || !ConfigPath.basicHomeExists() {
```
Please change this to

```
if !ConfigPath.xdgHomeExists() && ConfigPath.basicHomeExists() {
    return oldDefaultHelmHome
}
return defaultHelmHome
```

This makes it clear that using the oldDefaultHome is the exceptional case.

cleac replied:
Okay, I don't see a huge difference here

cleac committed:
Update docs and make code a bit more explicit; fix a lint

Signed-off-by: Alexander Nesterenko <nestorf250@gmail.com>

bacongobbler commented:
After reviewing the code, there's a *lot* of clean up required to bring this into shape before we can merge this. I'll close this PR out once I have a new one open with the changes necessary. Thank you @cleac for all your effort on this one!

bacongobbler commented:
closing in favour of https://github.com/helm/helm/pull/5443.
